File: src/app/model/HeartRateMetric.ts
import { MetricService } from "../services/metric.service";
import { HeartRateSensorServiceEvent } from "./HeartRateSensorServiceEvent";
import { Metric, MetricType } from "./Metric";

export class HeartRateMetric extends Metric {
    private handler = this.handleHeartRateEvent.bind(this);

    constructor(metricService: MetricService) {
        super(MetricType.HEART_RATE, 'Heart Rate', 'bpm', metricService);
    }

    startLogging(): void {
        this.metricService.getHeartRateSensorService().substribeForHeartRate(this.handler);
    }

    stopLogging(): void {
        this.metricService.getHeartRateSensorService().unsubscribeForHeartRate(this.handler);
    }

    handleHeartRateEvent(event: HeartRateSensorServiceEvent): void {
        this.addValue(event.heartRate.heartRate, event.heartRate.timestamp);
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/HeartRateMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnHeartRate;

public class HeartRateMetric extends Metric implements OnHeartRate {
    public HeartRateMetric(IMetricService metricService) {
        super(MetricType.HEART_RATE, "Heart Rate", "bpm", metricService);
    }

    @Override
    public void startLogging() {
        metricService.getHeartRateSensorService().subscribeForHeartRate(this);
    }

    @Override
    public void stopLogging() {
        metricService.getHeartRateSensorService().unsubscribeForHeartRate(this);
    }

    @Override
    public void onHeartRate(HeartRateData data) {
        addValue(data.heartRate, data.timestamp);
    }
}
